### PR TITLE
PS-3919: Assertion when analyze table performs a binlog rotation

### PR DESCRIPTION
**Pull request by dutow**
This commit:
* Adds a minimal testcase
* Backports the fix for this issue from 8.0, which can be found as part of two huge commits:
 * Disabling autocommit/analyze read lock for mysql_admin_table from 0a10fc8b5ca0
 * Disable_autocommit_guard RAII class from 31350e8ab151 (used by the otherr backported change)

### CODE REVIEW

File: mysql-test/t/percona_bug_ps3919-master.opt [hunk 1]
```diff
@@ -0,0 +1 @@
+--log_bin=binlog --gtid_mode=ON --enforce_gtid_consistency=ON --sql_mode=ONLY_FULL_GROUP_BY
```
Comment (laurynas-biveinis) [inline mysql-test/t/percona_bug_ps3919-master.opt]: Binlog suite?

Comment (laurynas-biveinis) [inline mysql-test/t/percona_bug_ps3919-master.opt]: Why does it need sql_mode?

File: sql/sql_admin.cc [hunk 2]
```diff
@@ -281,6 +281,13 @@ static bool mysql_admin_table(THD* thd, TABLE_LIST* tables,
                                                             HA_CHECK_OPT *),
                               int (view_operator_func)(THD *, TABLE_LIST*))
 {
+  /*
+    Prevent InnoDB from automatically committing InnoDB
+    transaction each time data-dictionary tables are closed after
```
Comment (laurynas-biveinis) [inline sql/sql_admin.cc]: This comment is not correct for 5.7

File: sql/sql_admin.cc [hunk 3]
```diff
@@ -1211,6 +1220,13 @@ bool Sql_cmd_analyze_table::execute(THD *thd)
   if (check_table_access(thd, SELECT_ACL | INSERT_ACL, first_table,
                          FALSE, UINT_MAX, FALSE))
     goto error;
+
+  DBUG_EXECUTE_IF("simulate_analyze_table_lock_wait_timeout_error",
```
Comment (laurynas-biveinis) [inline sql/sql_admin.cc]: Not used by the testcase

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]
Does the testcase pass on (either our or upstream) 8.0 branch as expected?

File: mysql-test/t/percona_bug_ps3919-master.opt (hunk 1 above)
Comment (dutow) [reply mysql-test/t/percona_bug_ps3919-master.opt]: Moved it there.

Comment (dutow) [reply mysql-test/t/percona_bug_ps3919-master.opt]: I'm not sure - but without that parameter, the crash didn't happen. The original bug report mentioned several other parameters, and I was able to remove most of them, but not this one.

File: sql/sql_admin.cc (hunk 2 above)
Comment (dutow) [reply sql/sql_admin.cc]: Removed it.

File: sql/sql_admin.cc (hunk 3 above)
Comment (dutow) [reply sql/sql_admin.cc]: Removed it.

**Comment by dutow** [comment]
For 8.0: yes, I tested the test case on our 8.0 branch, and it passed - I backported the changed from 8.0 after noticing that.

**Comment by dutow** [comment]
Jenkins: https://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1809/

**Review by laurynas-biveinis — APPROVED** [review]

**Pull request merged**